beetsplug/export.py

```diff
@@ -108,44 +135,118 @@ def run(self, lib, opts, args):
             except (mediafile.UnreadableFileError, IOError) as ex:
                 self._log.error(u'cannot read file: {0}', ex)
                 continue
-
             data = key_filter(data)
             items += [data]
 
-        export_format.export(items, **format_options)
+        self.run_results = items
+        self.export_format.export(self.run_results, **format_options)
 
 
 class ExportFormat(object):
     """The output format type"""
+    def __init__(self, file_path, file_mode=u'w', encoding=u'utf-8'):
+        self.path = file_path
+        self.mode = file_mode
+        self.encoding = encoding
+        # Used for testing
+        self.results = None
 
     @classmethod
-    def factory(cls, type, **kwargs):
-        if type == "json":
-            if kwargs['file_path']:
-                return JsonFileFormat(**kwargs)
-            else:
-                return JsonPrintFormat()
-        raise NotImplementedError()
+    def factory(cls, file_type, **kwargs):
+        if file_type == "json":
+            return JsonFormat(**kwargs)
+        elif file_type == "csv":
+            return CSVFormat(**kwargs)
+        elif file_type == "xml":
+            return XMLFormat(**kwargs)
+        else:
+            raise NotImplementedError()
 
     def export(self, data, **kwargs):
         raise NotImplementedError()
 
 
-class JsonPrintFormat(ExportFormat):
-    """Outputs to the console"""
+class JsonFormat(ExportFormat):
+    """Saves in a json file"""
+    def __init__(self, file_path, file_mode=u'w', encoding=u'utf-8'):
+        super(JsonFormat, self).__init__(file_path, file_mode, encoding)
+        self.export = self.export_to_file if self.path else self.export_to_terminal
+
+    def export_to_terminal(self, data, **kwargs):
+        r = json.dump(data, sys.stdout, cls=ExportEncoder, **kwargs)
+        self.results = str(r)
+
+    def export_to_file(self, data, **kwargs):
+        with codecs.open(self.path, self.mode, self.encoding) as f:
+            r = json.dump(data, f, cls=ExportEncoder, **kwargs)
+            self.results = str(r)
+
+
+class CSVFormat(ExportFormat):
+    """Saves in a csv file"""
+    def __init__(self, file_path, file_mode=u'w', encoding=u'utf-8'):
+        super(CSVFormat, self).__init__(file_path, file_mode, encoding)
+        self.header = []
 
     def export(self, data, **kwargs):
-        json.dump(data, sys.stdout, cls=ExportEncoder, **kwargs)
+        if data and type(data) is list and len(data) > 0:
```

Comment:
I'm a little confused by the `list` type check here—when will the passed object _not_ be a list? And what happens when it's not—so `self.header` remains empty?

Also, this condition could be simplified to:

```py
if data and isinstance(data, list):
```

because `data` is only "true" if it is non-empty.

Reply:
I was just trying to be extra cautious but I completely agree that it is a little overkill, this should work just fine instead; `if data and len(data) > 0:`